File: articles/epigenetics-regulation-gene-expression/fortran/epigenetics_kernel.f90
! Compact epigenetics and expression kinetics kernel in Fortran.

program epigenetics_kernel
  implicit none

  real :: m0
  real :: k
  real :: t
  real :: expr
  real :: half_life
  real :: methylated
  real :: unmethylated
  real :: meth_fraction
  real :: kon
  real :: koff
  real :: p_on_ss
  real :: treated_expr
  real :: control_expr
  real :: log2fc

  m0 = 120.0
  k = log(4.0) / 6.0
  t = 6.0

  expr = m0 * exp(-k * t)
  half_life = log(2.0) / k

  methylated = 85.0
  unmethylated = 15.0
  meth_fraction = methylated / (methylated + unmethylated)

  kon = 0.28
  koff = 0.10
  p_on_ss = kon / (kon + koff)

  treated_expr = 25.0
  control_expr = 12.0
  log2fc = log((treated_expr + 1.0e-6) / (control_expr + 1.0e-6)) / log(2.0)

  print *, "Expression at 6h:", expr
  print *, "Half-life:", half_life
  print *, "Methylation fraction:", meth_fraction
  print *, "Steady-state p_on:", p_on_ss
  print *, "log2FC:", log2fc
end program epigenetics_kernel

File: articles/epigenetics-regulation-gene-expression/c/epigenetics_kernel.c
/*
 * Compact epigenetics and gene-expression kernel in C.
 */

#include <math.h>
#include <stdio.h>

double transcript_decay(double m0, double k, double t) {
    return m0 * exp(-k * t);
}

double transcript_half_life(double k) {
    return log(2.0) / k;
}

double methylation_fraction(double methylated, double unmethylated) {
    double total = methylated + unmethylated;

    if (total <= 0.0) {
        return NAN;
    }

    return methylated / total;
}

double regulatory_steady_state(double kon, double koff) {
    return kon / (kon + koff);
}

int main(void) {
    double m0 = 120.0;
    double k = log(4.0) / 6.0;
    double log2fc = log((25.0 + 1e-6) / (12.0 + 1e-6)) / log(2.0);

    printf("expression_at_6h=%.6f\n", transcript_decay(m0, k, 6.0));
    printf("half_life_h=%.6f\n", transcript_half_life(k));
    printf("methylation_fraction=%.6f\n", methylation_fraction(85.0, 15.0));
    printf("p_on_steady_state=%.6f\n", regulatory_steady_state(0.28, 0.10));
    printf("log2FC_expr=%.6f\n", log2fc);

    return 0;
}
